tests/write_operations_test.go

```diff
@@ -0,0 +1,813 @@
+	fo, err := fio.Create(filePath)
+	s.Require().NoError(err)
+
+	s.Require().NoError(pqarrow.WriteTable(arrTbl, fo, arrTbl.NumRows(),
+		nil, pqarrow.DefaultWriterProps()))
+}
+
+func (s *WriteOperationsTestSuite) createTable(identifier table.Identifier, formatVersion int, spec iceberg.PartitionSpec) *table.Table {
+	meta, err := table.NewMetadata(s.tableSchema, &spec, table.UnsortedSortOrder,
+		s.location, iceberg.Properties{"format-version": strconv.Itoa(formatVersion)})
+	s.Require().NoError(err)
+
+	return table.New(
+		identifier,
+		meta,
+		s.getMetadataLoc(),
+		func(ctx context.Context) (iceio.IO, error) {
+			return iceio.LocalFS{}, nil
+		},
+		&mockedCatalog{},
+	)
+}
+
+func (s *WriteOperationsTestSuite) createTableWithData(identifier table.Identifier, numFiles int) (*table.Table, []string) {
+	tbl := s.createTable(identifier, 2, *iceberg.UnpartitionedSpec)
+	
+	files := make([]string, 0, numFiles)
+	fs := s.getFS(tbl)
+	
+	for i := 0; i < numFiles; i++ {
+		filePath := fmt.Sprintf("%s/data/test-%d.parquet", s.location, i)
+		s.writeParquet(fs, filePath, s.arrTable)
+		files = append(files, filePath)
+	}
+	
+	// Add files to table
+	tx := tbl.NewTransaction()
+	s.Require().NoError(tx.AddFiles(s.ctx, files, nil, false))
+	
+	committedTbl, err := tx.Commit(s.ctx)
+	s.Require().NoError(err)
+	
+	return committedTbl, files
+}
+
+func (s *WriteOperationsTestSuite) getFS(tbl *table.Table) iceio.WriteFileIO {
+	fs, err := tbl.FS(s.ctx)
+	s.Require().NoError(err)
+	return fs.(iceio.WriteFileIO)
+}
+
+func TestRewriteFiles(t *testing.T) {
+	suite.Run(t, &WriteOperationsTestSuite{})
+}
+
+func (s *WriteOperationsTestSuite) TestRewriteFiles() {
+	s.Run("RewriteDataFiles", func() {
+		// Setup table with multiple small files
+		ident := table.Identifier{"default", "rewrite_test_table"}
+		tbl, originalFiles := s.createTableWithData(ident, 3)
+		
+		// Create new consolidated file
+		consolidatedPath := fmt.Sprintf("%s/data/consolidated.parquet", s.location)
+		
+		// Create larger dataset for consolidation
+		mem := memory.DefaultAllocator
+		largerTable, err := array.TableFromJSON(mem, s.arrSchema, []string{
+			`[
+				{"id": 1, "data": "foo", "ts": 1672531200000000},
+				{"id": 2, "data": "bar", "ts": 1672534800000000},
+				{"id": 3, "data": "baz", "ts": 1672538400000000},
+				{"id": 4, "data": "qux", "ts": 1672542000000000},
+				{"id": 5, "data": "quux", "ts": 1672545600000000}
+			]`,
+		})
+		s.Require().NoError(err)
+		defer largerTable.Release()
+		
+		fs := s.getFS(tbl)
+		s.writeParquet(fs, consolidatedPath, largerTable)
+		
+		// Rewrite files (replace multiple small files with one larger file)
+		tx := tbl.NewTransaction()
+		err = tx.ReplaceDataFiles(s.ctx, originalFiles, []string{consolidatedPath}, nil)
+		s.Require().NoError(err)
+		
+		newTbl, err := tx.Commit(s.ctx)
+		s.Require().NoError(err)
+		
+		// Verify file count reduction
+		snapshot := newTbl.CurrentSnapshot()
```

Comment:
is it worth to add a check for snapshot state? (i.e. - check that files are the one that you expect to see)

Reply:
added in https://github.com/apache/iceberg-go/pull/497/commits/7210f86e198f22a24736a368f46c7edfd161e536